pkg/controller/operators/labeller/labels.go

```diff
@@ -71,58 +93,77 @@ func ObjectPatchLabeler(
 	ctx context.Context,
 	logger *logrus.Logger,
 	check func(metav1.Object) bool,
+	list func(selector labels.Selector) (ret []*metav1.PartialObjectMetadata, err error),
 	patch func(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *apiextensionsv1.CustomResourceDefinition, err error),
-) func(
-	obj interface{},
-) error {
-	return func(obj interface{}) error {
-		cast, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
-		if !ok {
-			err := fmt.Errorf("wrong type %T, expected %T: %#v", obj, new(*apiextensionsv1.CustomResourceDefinition), obj)
-			logger.WithError(err).Error("casting failed")
-			return fmt.Errorf("casting failed: %w", err)
-		}
+) func(done func() bool) queueinformer.LegacySyncHandler {
+	return func(done func() bool) queueinformer.LegacySyncHandler {
+		return func(obj interface{}) error {
+			cast, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+			if !ok {
+				err := fmt.Errorf("wrong type %T, expected %T: %#v", obj, new(*apiextensionsv1.CustomResourceDefinition), obj)
+				logger.WithError(err).Error("casting failed")
+				return fmt.Errorf("casting failed: %w", err)
+			}
 
-		if !check(cast) || hasLabel(cast) {
-			return nil
-		}
+			if !check(cast) || hasLabel(cast) {
+				// if the object we're processing does not need us to label it, it's possible that every object that requires
+				// the label already has it; in which case we should exit the process, so the Pod that succeeds us can filter
+				// the informers used to drive the controller and stop having to track extraneous objects
+				items, err := list(labels.Everything())
+				if err != nil {
+					logger.WithError(err).Warn("failed to list all objects to check for labelling completion")
+					return nil
+				}
+				gvrFullyLabelled := true
+				for _, item := range items {
+					gvrFullyLabelled = gvrFullyLabelled && (!check(item) || hasLabel(item))
+				}
+				if gvrFullyLabelled {
+					allObjectsLabelled := done()
+					if allObjectsLabelled {
+						logrus.Fatal("detected that every object is labelled, exiting...")
+					}
+				}
+				return nil
+			}
 
-		uid := cast.GetUID()
-		rv := cast.GetResourceVersion()
+			uid := cast.GetUID()
```

Comment:
is it worth moving this step into its own function just to reduce the footprint of this function a bit?

Reply:
this = 130 - 165

Reply:
If possible, we prefer not to touch this code - the PR I made upstream to get applyconfigurations created for CRDs is in, and backported, so the next time we bump k8s deps we can delete this function